devito/operations/solve.py

```diff
@@ -0,0 +1,70 @@
+from functools import singledispatch
+
+from devito.finite_differences.differentiable import Add, Mul, EvalDerivative
+
+__all__ = ['linsolve', 'SolveError']
+
+
+class SolveError(Exception):
+    """Base class for exceptions in this module."""
+    pass
+
+
+def linsolve(expr, target, **kwargs):
+    """
+    Linear solve for the targe in a single equation.
+
+    Parameters
+    ----------
+    expr : expr-like
+        The expr to be rearranged.
+    target : symbol
+        The symbol w.r.t. which the equation is rearranged. May be a `Function`
+        or any other symbolic object.
+    """
+    c = factorize_target(expr, target)
+    if c != 0:
+        return -expr.xreplace({target: 0})/c
+    raise SolveError("No linear solution found")
+
+
+def eval_t(expr):
+    """
+    Evaluate all time derivatives in the expression.
+    """
+    try:
+        assert any(d.is_Time for d in expr.dims)
+        return expr.evaluate
+    except:
```

Comment:
I know this is what we have in master, but I can't really see this nested try-except with no exception specification. What exceptions should be caught here? Also, always good to de-nest the exceptions when possible (I think it is here, just add `pass` in the first one and de-nest)

EDIT: oh, the except is for the assert? but that's not how `assert` should be used... 

```
if any(d.is_Time for d in expr.dims):
    return expr.evaluate
else:
    try:
        return ...
    except X:
        return expr
```

how about above ?

Reply:
Well no that won't work because `expr` may not have a `dims` od `d` may not have a `is_Time`.

Reply:
reworked with dispatch :D another X2 speedup in the process.